pkg/loop/telem.go

```diff
@@ -2,34 +2,118 @@ package loop
 
 import (
 	"context"
+	"os"
+	"runtime/debug"
+	"time"
 
 	grpcprom "github.com/grpc-ecosystem/go-grpc-middleware/providers/prometheus"
 	"github.com/prometheus/client_golang/prometheus"
 	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
 	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
 	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
 	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
 	"go.opentelemetry.io/otel/trace"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
 
 	"github.com/smartcontractkit/chainlink-relay/pkg/loop/internal"
 )
 
 type GRPCOpts = internal.GRPCOpts
 
-// SetupTelemetry initializes open telemetry and returns GRPCOpts with telemetry interceptors.
-func SetupTelemetry(registerer prometheus.Registerer) GRPCOpts {
-	otel.SetTracerProvider(sdktrace.NewTracerProvider(
-		sdktrace.WithSampler(sdktrace.AlwaysSample()),
-	))
-	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
+type TracingConfig struct {
+	// NodeAttributes are the attributes to attach to traces.
+	NodeAttributes map[string]string
+
+	// Enables tracing; requires a collector to be provided
+	Enabled bool
 
+	// Collector is the address of the OTEL collector to send traces to.
+	CollectorTarget string
+
+	// SamplingRatio is the ratio of traces to sample. 1.0 means sample all traces.
+	SamplingRatio float64
+}
+
+// NewGRPCOpts initializes open telemetry and returns GRPCOpts with telemetry interceptors.
+// It is called from the host and each plugin - intended as there is bidirectional communication
+func NewGRPCOpts(registerer prometheus.Registerer) GRPCOpts {
 	if registerer == nil {
 		registerer = prometheus.DefaultRegisterer
 	}
 	return GRPCOpts{DialOpts: dialOptions(registerer), NewServer: newServerFn(registerer)}
 }
 
+// SetupTracing initializes open telemetry with the provided config.
+// It sets the global trace provider and opens a connection to the configured collector.
+func SetupTracing(config TracingConfig) error {
+	if !config.Enabled {
+		return nil
+	}
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+
+	// Enough to shutdown the underlying connection since DialContext is used in blocking mode
+	defer cancel()
+	conn, err := grpc.DialContext(ctx, config.CollectorTarget,
+		// Note the use of insecure transport here. TLS is recommended in production.
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
```

Comment:
This is a pretty big caveat. Any reason not to wire it up to config now? Seems worth noting in the func doc and creating an issue for.

Reply:
It is - I tabled it for now because the otel collector is always living directly next to the container for node (whether locally or in PRs). Agreed on doc/issue.